File: frontend/src/types/graph.ts
// Mirrors backend/app/models/graph.py exactly (hand-kept in sync -- no
// codegen step for this app). Field names are snake_case to match the
// Pydantic JSON wire format directly, with no transformation either side.

export type NodeType =
  | "room"
  | "corridor"
  | "door"
  | "stair"
  | "elevator"
  | "entrance"
  | "outdoor";

export type EdgeType = "hallway" | "door" | "stairs" | "elevator" | "outdoor_path";

// Named GraphNode/GraphEdge, not Node/Edge, to avoid colliding with the DOM's
// global `Node` type.
export interface GraphNode {
  id: string;
  building: string;
  floor: number;
  x: number;
  y: number;
  type: NodeType;
  label: string | null;
  room_type: string | null;
  department: string | null;
}

export interface GraphEdge {
  id: string;
  from_node: string;
  to_node: string;
  weight: number;
  type: EdgeType;
  points?: [number, number][];
}

export interface PassageLine {
  id: string;
  points: [number, number][];
}

export interface DoorPassageAttachment {
  door_node_id: string;
  pathway_point: [number, number];
  points: [number, number][];
}

export type RoutingSource =
  | "legacy"
  | "anchors_only"
  | "gemini_pathways"
  | "space_type_centerlines";

/**
 * Whether a graph stores its corridors as passage polylines (rather than as
 * persisted corridor nodes). Both the geometric centerline builder and the
 * older vision tracer produce that shape, so the UI treats them the same:
 * it draws passage lines and allows routing.
 */
export function usesPassageLines(source: RoutingSource | null | undefined): boolean {
  return source === "gemini_pathways" || source === "space_type_centerlines";
}

export interface FloorGraph {
  building: string;
  floor: number;
  nodes: GraphNode[];
  edges: GraphEdge[];
  auto_generated: boolean;
  routing_source: RoutingSource;
  page_width: number;
  page_height: number;
  routing_warnings: string[];
  unconnected_room_ids: string[];
  passageways: PassageLine[];
  door_attachments: DoorPassageAttachment[];
}

File: backend/app/models/graph.py
from typing import Literal, Optional

from pydantic import BaseModel

NodeType = Literal[
    "room", "corridor", "door", "stair", "elevator", "entrance", "outdoor"
]
EdgeType = Literal["hallway", "door", "stairs", "elevator", "outdoor_path"]


class Node(BaseModel):
    id: str
    building: str
    floor: int
    x: float
    y: float
    type: NodeType
    label: Optional[str] = None
    room_type: Optional[str] = None
    department: Optional[str] = None


class Edge(BaseModel):
    id: str
    from_node: str
    to_node: str
    weight: float
    type: EdgeType
    points: list[tuple[float, float]] = []


class PassageLine(BaseModel):
    id: str
    points: list[tuple[float, float]]


class DoorPassageAttachment(BaseModel):
    door_node_id: str
    pathway_point: tuple[float, float]
    points: list[tuple[float, float]]


class FloorGraph(BaseModel):
    building: str
    floor: int
    nodes: list[Node] = []
    edges: list[Edge] = []
    auto_generated: bool = False
    routing_source: Literal[
        "legacy", "anchors_only", "gemini_pathways", "space_type_centerlines"
    ] = "legacy"
    page_width: float = 1224.0
    page_height: float = 792.0
    routing_warnings: list[str] = []
    unconnected_room_ids: list[str] = []
    passageways: list[PassageLine] = []
    door_attachments: list[DoorPassageAttachment] = []


class WallSegment(BaseModel):
    type: Literal["line", "curve"]
    points: list[tuple[float, float]]
    layer: Optional[str] = None  # CAD layer name (e.g. "ARCH|A-WALL"), when the PDF preserves it


class FloorDraft(BaseModel):
    """Output of the deterministic extraction pipeline: room nodes plus raw
    vector geometry, before a human has drawn any corridor/door edges."""

    building: str
    floor: int
    nodes: list[Node] = []
    raw_geometry: list[WallSegment] = []
    # Closed per-room boundaries from the CAD "A-AREA" layer, when the PDF
    # carries them. Kept as whole polygons rather than loose segments because
    # the grouping is the useful part: it answers which room a point is in,
    # and therefore which room a door actually belongs to.
    room_polygons: list[list[tuple[float, float]]] = []
    # Node id -> the doorways of that space, found from where its outline
    # opens onto walkable floor (see services/space_doors.py). Attributed to
    # the space by construction, so a door is never claimed by the room on
    # the other side of the wall.
    space_doors: dict[str, list[tuple[float, float]]] = {}
    raster_path: Optional[str] = None
